File: src/main/scala/Maintenance.scala
package com.tcroonen

import org.apache.iceberg.Table
import org.apache.iceberg.spark.actions.SparkActions
import org.apache.spark.sql.SparkSession

object Maintenance {

  def expireSnapshots(spark: SparkSession,table: Table): Unit = {
    SparkActions.get(spark)
      .expireSnapshots(table)
      .expireOlderThan(System.currentTimeMillis())
      .retainLast(1)
      .execute
  }

  def rewriteDataFiles(spark: SparkSession, table: Table): Unit = {
    SparkActions.get(spark)
      .rewriteDataFiles(table)
      .option("target-file-size-bytes", (1024 * 1024 * 100L).toString)
      .binPack
      .execute
  }

  def deleteOrphanFiles(spark: SparkSession, table: Table): Unit = {
    SparkActions.get(spark)
      .deleteOrphanFiles(table)
      .olderThan(System.currentTimeMillis() - 1000L*60*60*24*7)
      .execute
  }

  def rewriteManifests(spark: SparkSession, table: Table): Unit = {
    SparkActions.get(spark)
      .rewriteManifests(table)
      .execute
  }
}

File: python/src/maintenance.py
from datetime import datetime, timedelta

from pyspark.sql import SparkSession


def rewrite_data_files(spark: SparkSession, catalog: str, table: str) -> None:
    spark.sql(
        f"""
        CALL {catalog}.system.rewrite_data_files(
            table => '{table}', 
            strategy => 'binpack', 
            options => map('target-file-size-bytes', {1024 * 1024 * 100})
        )
        """
    )


def expire_snapshots(spark: SparkSession, catalog: str, table: str) -> None:
    spark.sql(
        f"""
        CALL {catalog}.system.expire_snapshots(
            table => '{table}', 
            older_than => TIMESTAMP '{datetime.now().strftime("%Y-%m-%d %H:%M:%S")}', 
            retain_last => 1
        )
        """
    )


def delete_orphan_files(spark: SparkSession, catalog: str, table: str) -> None:
    ts = (datetime.now() - timedelta(days=7)).strftime("%Y-%m-%d %H:%M:%S")

    spark.sql(
        f"""
        CALL {catalog}.system.remove_orphan_files(
            table => '{table}', 
            older_than => TIMESTAMP '{ts}'
        )
        """
    )


def rewrite_manifests(spark: SparkSession, catalog: str, table: str) -> None:
    spark.sql(
        f"""
        CALL {catalog}.system.rewrite_manifests(
            table => '{table}'
        )
        """
    )
